README.md
<p align="center" >
  <img src="https://raw.githubusercontent.com/malcommac/SwiftMsgPack/develop/logo.png" width=210px height=204px alt="SwiftMsgPack" title="SwiftMsgPack">
</p>

[![Carthage compatible](https://img.shields.io/badge/Carthage-compatible-4BC51D.svg?style=flat)](https://github.com/Carthage/Carthage) [![CI Status](https://travis-ci.org/malcommac/SwiftMsgPack.svg)](https://travis-ci.org/malcommac/SwiftMsgPack) [![Version](https://img.shields.io/cocoapods/v/SwiftMsgPack.svg?style=flat)](http://cocoadocs.org/docsets/SwiftMsgPack) [![License](https://img.shields.io/cocoapods/l/SwiftMsgPack.svg?style=flat)](http://cocoadocs.org/docsets/SwiftMsgPack) [![Platform](https://img.shields.io/cocoapods/p/SwiftMsgPack.svg?style=flat)](http://cocoadocs.org/docsets/SwiftMsgPack)

<p align="center" >MsgPack for Swift... Its' like JSON <a href="http://theburningmonk.com/2011/12/performance-test-binary-serializers-part-ii/"> but faster</a><br/>
Made with ♥ in pure Swift, no dependencies, lightweight & fully portable
<p/>
<p align="center" >★★ <b>Star our github repository to help us!</b> ★★</p>
<p align="center" >Created by <a href="http://www.danielemargutti.com">Daniele Margutti</a> (<a href="http://www.twitter.com/danielemargutti">@danielemargutti</a>)</p>

[MessagePack](http://msgpack.org/) is an efficient binary serialization format, which lets you exchange data among multiple languages like JSON, except that it's faster and smaller. Small integers are encoded into a single byte while typical short strings require only one extra byte in addition to the strings themselves.
You can read more about [specs directly from the main web](https://github.com/msgpack/msgpack/blob/master/spec.md) site.

## You also may like

Do you like `SwiftMsgPack`? I'm also working on several other opensource libraries.

Take a look here:

* **[SwiftDate](https://github.com/malcommac/SwiftDate)** - Date & Timezone management in Swift
* **[Hydra](https://github.com/malcommac/Hydra)** - Promise, Async/Await on sterioids!
* **[SwiftLocation](https://github.com/malcommac/SwiftLocation)** - CoreLocation and Beacon Monitoring on steroid!
* **[SwiftRichString](https://github.com/malcommac/SwiftRichString)** - Elegant and painless attributed string in Swift
* **[SwiftScanner](https://github.com/malcommac/SwiftScanner)** - String scanner in pure Swift with full unicode support
* **[SwiftSimplify](https://github.com/malcommac/SwiftSimplify)** - Tiny high-performance Swift Polyline Simplification Library

## Index
* **[Current Release](#release)**
* **[How to use](#howto)**
* **[Supported Types](#supportedtypes)**
* **[Installation (CocoaPods, SwiftPM and Carthage)](#installation)**
* **[Tests](#tests)**
* **[Requirements](#requirements)**
* **[Credits](#credits)**

<a name="release" />

## Current Release

* **Swift 4.x**: Latest is 1.0.0 [Download here](https://github.com/malcommac/SwiftMsgPack/releases/tag/1.0.0).
* **Swift 3.x**: Last is 0.9.1 [Download here](https://github.com/malcommac/SwiftMsgPack/releases/tag/0.9.1).

A complete list of changes for each release is available in the [CHANGELOG](CHANGELOG.md) file.

<a name="howto" />

## How to use: one-shot pack & unpack

Both serialization and deseralization happens inside a `Data` object.

In order to pack one or more objects you need to create an empty `Data` instance and call `pack` by passing the object(s) you want to serialize.
It's pretty easy, take a look here:

```swift
var data = Data()
do {
	let obj1 = "Hello World"
	let obj2 = 45.5
	let obj3: [AnyHashable:Any?] = [ "key_1" : "value test","key_2" : 4,"key_3" : true, otherHashableKey: "value1"]
	// Now you can pack your instances by passing them to pack function
	try data.pack(obj1,obj2,obj3)
} catch {
	print("Something went wrong while packing data: \(error)")	
}
```
Deserializing data is pretty simple too, just call `unpack` function to an instance of `Data` with msgpack data:

```swift
let data: Data = // msgpack data...
do {
  let decodedObj: Any? = try data.unpack()
} catch {
	print("Something went wrong while unpacking data: \(error)")	
}
```

<a name="supportedtypes" />

## Supported Types

SwiftMsgPack supports the following Swift types:
- `String`
- `Data`
- `Bool`
- `nil`
- Numeric values: `Int` & `UInt` (`UInt8`,`Int8`,`UInt16`,`Int16`,`UInt32`,`Int32`,`UInt64`,`Int64`), `Float` and `Double`
- `Dictionaries` (`[AnyHashable:Any?]`)
- `Array` (`[Any?]`)

The following limitations are specified by [MsgPack specs format](https://github.com/msgpack/msgpack/blob/master/spec.md#types-limitation):

* a value of an Integer object is limited from `-(2^63)` upto `(2^64)-1`
* maximum length of a Binary object is `(2^32)-1`
* maximum byte size of a String object is `(2^32)-1`
* String objects may contain invalid byte sequence and the behavior of a deserializer depends on the actual implementation when it received invalid byte sequence
    * Deserializers should provide functionality to get the original byte array so that applications can decide how to handle the object
* maximum number of elements of an Array object is `(2^32)-1`
* maximum number of key-value associations of a Map object is `(2^32)-1`

<a name="installation" />

## Installation
You can install Swiftline using CocoaPods, carthage and Swift package manager

### CocoaPods
    use_frameworks!
    pod 'SwiftMsgPack'

### Carthage
    github 'malcommac/SwiftMsgPack'

### Swift Package Manager
Add swiftline as dependency in your `Package.swift`

```
  import PackageDescription

  let package = Package(name: "YourPackage",
    dependencies: [
      .Package(url: "https://github.com/malcommac/SwiftMsgPack.git", majorVersion: 0),
    ]
  )
```

<a name="tests" />

## Tests

SwiftMsgPack has an extensive coverage using XCTest.
You can found a complete list of tests inside `Tests/SwiftMsgPackTests` folder.
Tests can also be runned with XCode using the SwiftMsgPack project.

<a name="requirements" />

## Requirements

Current version is compatible with:

* Swift 4.x+ (Swift 3.x up to 0.9.1)
* iOS 9.0 or later
* tvOS 9.0 or later
* macOS 10.10 or later
* Linux compatible environments

<a name="credits" />

## Credits & License
SwiftMsgPack is owned and maintained by [Daniele Margutti](http://www.danielemargutti.com/).

As open source creation any help is welcome!

The code of this library is licensed under MIT License; you can use it in commercial products without any limitation.

The only requirement is to add a line in your Credits/About section with the text below:

```
This software uses open source SwiftMsgPack's library to manage MsgPack data.
Web: http://github.com/malcommac/SwiftMsgPack.
Created by Daniele Margutti and licensed under MIT License.
```

msgpack.org[SwiftMsgPack]


CHANGELOG.md
<p align="center" >
  <img src="https://raw.githubusercontent.com/malcommac/SwiftMsgPack/develop/logo.png" width=210px height=204px alt="SwiftMsgPack" title="SwiftMsgPack">
</p>

## CHANGELOG

* Version **[1.0.0](#100)**
* Version **[0.9.0](#090)**

<a name="100" />

## SwiftMsgPack 1.0.0
---
- **Release Date**: 2017/09/14
- **Zipped Version**: [Download 1.0.0](https://github.com/malcommac/SwiftMsgPack/releases/tag/1.0.0)

Swift 4 compatible version

<a name="090" />

## SwiftMsgPack 0.9.0
---
- **Release Date**: 2017/02/26
- **Zipped Version**: [Download 0.9.0](https://github.com/malcommac/SwiftMsgPack/releases/tag/0.9.0)

First release


Tests/SwiftMsgPackTests/SwiftMsgPackTests_Data.swift
/*
* SwiftMsgPack
* Lightweight MsgPack for Swift
*
* Created by:	Daniele Margutti
* Email:			hello@danielemargutti.com
* Web:			http://www.danielemargutti.com
* Twitter:		@danielemargutti
*
* Copyright © 2017 Daniele Margutti
*
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the "Software"), to deal
* in the Software without restriction, including without limitation the rights
* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
* copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in
* all copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
* THE SOFTWARE.
*
*/

import Foundation

import XCTest
@testable import SwiftMsgPack

class SwiftMsgPackTests_Data: XCTestCase {
	
	override func setUp() {
		super.setUp()
	}
	
	override func tearDown() {
		super.tearDown()
	}
	
	// MARK: - Test Data

	func test_shortData() {
		let len = 6
		let (data,exp) = generateTestData(len)
		performTestData(name: "Test Short Data", value: data, expected: exp)
	}
	
	func test_mediumData() {
		let len = Int( UInt16(UInt16.max - 1) )
		let (data,exp) = generateTestData(len)
		performTestData(name: "Test Medium Data", value: data, expected: exp)
	}
	
	func test_bigData() {
		let len = Int( UInt32(UInt16.max) + 1 )
		let (data,exp) = generateTestData(len)
		performTestData(name: "Test Long Data", value: data, expected: exp)
	}
	
	// MARK: - Helper Functions
	
	func generateTestData(_ length: Int) -> (Data,[UInt8]) {
		var bytes: [UInt8] = []
		let data_sequence: [UInt8] = generateRandomNumberSequence(length)
		
		if length < Int(UInt8.max) {
			// Header
			bytes.append(UInt8(0xc4))
			// Length
			bytes.append(UInt8(length))
		}
		else if length < Int(UInt16.max) {
			// Header
			bytes.append(UInt8(0xc5))
			// Length (big endian)
			bytes.append(UInt8((length >> 8) & 0xff))
			bytes.append(UInt8(length & 0xff))
		}
		else if length < Int(UInt32.max) {
			bytes.append(UInt8(0xc6))
			// Length (big endian)
			bytes.append(UInt8((length >> 24) & 0xff))
			bytes.append(UInt8((length >> 16) & 0xff))
			bytes.append(UInt8((length >> 8) & 0xff))
			bytes.append(UInt8(length & 0xff))
		}
		// Append real data
		bytes.append(contentsOf: data_sequence)
		// Get generated data
		let data = Data(bytes: data_sequence)
		return (data,bytes)
	}
	
	func generateRandomNumberSequence(_ length: Int) -> [UInt8] {
		var items: [UInt8] = []
		for _ in 0..<length {
			let val = ( UInt8.max - UInt8.min) + UInt8.min
			let randomNum = Int(arc4random_uniform(UInt32(val)))
			items.append(UInt8(randomNum))
		}
		return items
	}
	
	func performTestData(name testName: String, value: Data, expected bytes: [UInt8]) {
		var packed = Data()
		
		do {
			try packed.pack(value)
			
			// Check resulting data size
			guard bytes.count == packed.count else {
				XCTFail("[\(testName)] Resulting packed data is different in byte size than expected (\(packed.count), \(bytes.count) expected)")
				return
			}
			
			// Validate each byte
			var idx = 0
			for byte in packed {
				guard byte == bytes[idx] else {
					XCTFail("[\(testName)] Byte \(idx) is different from expected (\(byte), \(bytes[idx]) expected)")
					return
				}
				idx += 1
			}
			
			// Unpack data
			guard let unpacked = try packed.unpack() else {
				XCTFail("[\(testName)] Failed to unpack data")
				return
			}
			
			// Cast to Data object
			guard let unpacked_data = unpacked as? Data else {
				XCTFail("[\(testName)] Failed to cast unpacked data to `Data` instance")
				return
			}
			
			// Check if data is equal
			guard unpacked_data == value else {
				XCTFail("[\(testName)] Unpackaed data is different from expected")
				return
			}
			
		} catch let err {
			// Something went wrong while packing data
			XCTFail("[\(testName)] Failed to pack data: \(err) (src='\(value)')")
			return
		}
	}
	
}


Tests/SwiftMsgPackTests/SwiftMsgPackTests_Array.swift
/*
* SwiftMsgPack
* Lightweight MsgPack for Swift
*
* Created by:	Daniele Margutti
* Email:			hello@danielemargutti.com
* Web:			http://www.danielemargutti.com
* Twitter:		@danielemargutti
*
* Copyright © 2017 Daniele Margutti
*
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the "Software"), to deal
* in the Software without restriction, including without limitation the rights
* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
* copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in
* all copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
* THE SOFTWARE.
*
*/


import XCTest
@testable import SwiftMsgPack

class SwiftMsgPackTests_Array: XCTestCase {
	
	override func setUp() {
		super.setUp()
	}
	
	override func tearDown() {
		super.tearDown()
	}
	
	// MARK: - Test on Array type
	
	
	/// Test small array
	func testSmallArray() {
		let len = 15
		let (array,exp) = generateTestArray(len)
		performTestOnArray(name: "Test Small Array", value: array, expected: exp)
	}
	
	/// Test medium array
	func testMediumArray() {
		let len = UInt16(UInt16.max - 1)
		let (array,exp) = generateTestArray(Int(len))
		performTestOnArray(name: "Test Medium Array", value: array, expected: exp)
	}
	
	/// Test long array
	func testLongArray() {
		let len = UInt32(UInt16.max) + 1
		let (array,exp) = generateTestArray(Int(len))
		performTestOnArray(name: "Test Long Array", value: array, expected: exp)
	}
	
	// MARK: Helper Functions

	/// Generate a Test array and return both it and expected bytes for its packed version
	///
	/// - Parameter count: number of items in array
	/// - Returns: array and expected bytes
	func generateTestArray(_ count: Int) -> ([Any?],[UInt8]) {
		var array: [Any?] = []
		var bytes: [UInt8] = []
		
		// Generate Header for Array
		if count < 16 {
			// Header + Length
			bytes.append(UInt8(0x90 + count))
		}
		else if count < Int(UInt16.max) {
			// Header
			bytes.append(UInt8(0xdc))
			// Length (big endian)
			bytes.append(UInt8((count >> 8) & 0xff))
			bytes.append(UInt8(count & 0xff))
		}
		else if count < Int(UInt32.max) {
			// Header
			bytes.append(UInt8(0xdd))
			// Length (big endian)
			bytes.append(UInt8((count >> 24) & 0xff))
			bytes.append(UInt8((count >> 16) & 0xff))
			bytes.append(UInt8((count >> 8) & 0xff))
			bytes.append(UInt8(count & 0xff))
		}
		
		// Append items to array
		for _ in 0..<count {
			let (generatedItem,itemBytes) = generateRandomArrayType()
			// Append item to our array
			array.append(generatedItem)
			// Add stream to the bytes array
			bytes.append(contentsOf: itemBytes)
		}
		
		return (array,bytes)
	}
	
	
	/// Generate random type and return bytes stream which describe it
	///
	/// - Returns: item, byte which describe it
	func generateRandomArrayType() -> (Any?,[UInt8]) {
		let type = randomValue(min: 0, max: 3)
		var item: Any?
		var bytes: [UInt8] = []
		
		// Just to test some types
		switch type {
		case 0: // nil
			item = nil
			
			let header: UInt8 = UInt8(0xc0)
			bytes.append(header)
		case 1: // Boolean
			let bool_val = randomValue(min: 0, max: 1)
			item = (bool_val == 0 ? false : true)
			
			let header: UInt8 = (item as! Bool == true ? UInt8(0xc3) : UInt8(0xc2))
			bytes.append(header)
		case 2: // Int
			let int8Num = randomValue(min: 0, max: Int(Int8.max))
			item = int8Num
			
			let header: UInt8 = UInt8(int8Num & 0x7f)
			bytes.append(header)
		case 3: // Fixed String
			let random_len = randomValue(min: 1, max: 20)
			let str = randomString(length: random_len)
			item = str
			
			let header: UInt8 = UInt8(UInt8(0xa0) + UInt8(random_len))
			bytes.append(header)
			
			let encoded_data = str.data(using: String.Encoding.utf8)!
			bytes.append(contentsOf: [UInt8](encoded_data))
		default:
			break
		}

		return (item,bytes)
	}
	
	/// Generate a random string with passed length
	///
	/// - Parameter length: length of the string
	/// - Returns: random alphanumeric string
	func randomString(length: Int) -> String {
		let letters : NSString = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
		let len = UInt32(letters.length)
		var randomString = ""
		for _ in 0 ..< length {
			let rand = arc4random_uniform(len)
			var nextChar = letters.character(at: Int(rand))
			randomString += NSString(characters: &nextChar, length: 1) as String
		}
		return randomString
	}
	
	
	/// Generate a random number in a range
	///
	/// - Parameters:
	///   - min: min value
	///   - max: max value
	/// - Returns: random value
	func randomValue(min: Int, max: Int) -> Int {
		return Int(arc4random_uniform(UInt32(max) - UInt32(min)) + UInt32(min))
	}
	
	func performTestOnArray(name testName: String, value: [Any?], expected bytes: [UInt8]) {
		var packed = Data()
		
		do {
			try packed.pack(value)
			
			guard packed.count == bytes.count else {
				XCTFail("[\(testName)] Pack failed because number of bytes is different from expected: \(packed.count), \(bytes.count) expected")
				return
			}
			
			var idx = 0
			for byte in packed {
				guard byte == bytes[idx] else {
					XCTFail("[\(testName)] Byte \(idx) is different from expected (\(byte), \(bytes[idx]) expected")
					return
				}
				idx += 1
			}
			
			
		} catch let err {
			// Something went wrong while packing data
			XCTFail("[\(testName)] Failed to pack Array: \(err) (src='\(value)')")
			return
		}
	}
}


SwiftMsgPack/SwiftMsgPackTests/SwiftMsgPackTests.swift
//
//  SwiftMsgPackTests.swift
//  SwiftMsgPackTests
//
//  Created by Daniele Margutti on 26/02/2017.
//  Copyright © 2017 moka. All rights reserved.
//

import XCTest
@testable import SwiftMsgPack

class SwiftMsgPackTests: XCTestCase {
    
    override func setUp() {
        super.setUp()
        // Put setup code here. This method is called before the invocation of each test method in the class.
    }
    
    override func tearDown() {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
        super.tearDown()
    }
    
    func testExample() {
        // This is an example of a functional test case.
        // Use XCTAssert and related functions to verify your tests produce the correct results.
    }
    
    func testPerformanceExample() {
        // This is an example of a performance test case.
        self.measure {
            // Put the code you want to measure the time of here.
        }
    }
    
}


Sources/SwiftMsgPack/Decoder.swift
/*
* SwiftMsgPack
* Lightweight MsgPack for Swift
*
* Created by:	Daniele Margutti
* Email:			hello@danielemargutti.com
* Web:			http://www.danielemargutti.com
* Twitter:		@danielemargutti
*
* Copyright © 2017 Daniele Margutti
*
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the "Software"), to deal
* in the Software without restriction, including without limitation the rights
* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
* copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in
* all copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
* THE SOFTWARE.
*
*/

import Foundation

// MARK: Helper Struct to read and decoded stream of `Data`

private struct StreamReader {
	
	/// Pointer to `data` instance
	private var data: Data
	
	/// Current position on `data` instance
	private var index: Int
	
	/// Initialize a new stream with an input data
	///
	/// - Parameter data: data to read
	public init(_ data: Data) {
		self.data = data
		self.index = 0
	}
	
	/// Read the next data type header of the structure (8 bit) and move the index by 1 position
	///
	/// - Returns: data type value
	/// - Throws: throw an exception if data cannot be read or it's not available
	mutating func readType() throws -> UInt8 {
		guard index < data.count else {
			throw MsgPackError.unexpectedData
		}
		let type: UInt8 = data[Data.Index(index)]
		index += 1
		return type
	}
	
	/// Read 8 bit value and return it and move the index by 1 position
	///
	/// - Returns: 8 bit value
	/// - Throws: throw an exception if data cannot be read or it's not available
	mutating func read8Bit() throws -> UInt8 {
		return try readType()
	}
	
	
	/// Read next 16 bytes and return the value (index is moved according to the task)
	///
	/// - Returns: value read
	/// - Throws: throw an exception if data cannot be read or it's not available
	mutating func read16Bit() throws -> UInt16 {
		guard index + 2 <= data.count else {
			throw MsgPackError.unexpectedData
		}
		let value_int16 = UInt16(data[Data.Index(index)]) << 8 + UInt16(data[Data.Index(index + 1)])
		index += 2
		return value_int16
	}
	
	/// Read next 32 bytes and return the value (index is moved according to the task)
	///
	/// - Returns: value read
	/// - Throws: throw an exception if data cannot be read or it's not available
	mutating func read32Bit() throws -> UInt32 {
		guard index + 4 <= data.count else {
			throw MsgPackError.unexpectedData
		}
		var value_int32: UInt32 = 0
		for idx in index...(index + 3) {
			value_int32 = (value_int32 << 8) + UInt32(data[Data.Index(idx)])
		}
		index += 4
		return value_int32
	}
	
	/// Read next 64 bytes and return the value (index is moved according to the task)
	///
	/// - Returns: value read
	/// - Throws: throw an exception if data cannot be read or it's not available
	mutating func read64Bit() throws -> UInt64 {
		guard index + 8 <= data.count else {
			throw MsgPackError.unexpectedData
		}
		var value_int64: UInt64 = 0
		for idx in index...(index + 7) {
			value_int64 = (value_int64 << 8) + UInt64(data[Data.Index(idx)])
		}
		index += 8
		return value_int64
	}
	
	
	/// Read next `length` bytes of data and return it (index is moved according to it)
	///
	/// - Parameter length: length of data to read
	/// - Returns: read value
	/// - Throws: throw an exception if data cannot be read or it's not available
	mutating func readData(length: Int) throws -> Data {
		guard index + length <= data.count else {
			throw MsgPackError.unexpectedData
		}
		let range = Range(index..<(index + length))
		index += length
		return data.subdata(in: range)
	}
}

// MARK: Unpack of MsgPack Data

public extension Data {
	
	// MARK - Unpack
	
	/// This is the public function which can read a sequence of Data
	/// and unpack all objects by returning decoded data.
	///
	/// - Returns: decoded data
	/// - Throws: an error if decoding task cannot be finished correctly due to an error
	public func unpack() throws -> Any? {
		// Create a reader which has a point to the current position in data instance
		// and several help functions to read data
		var reader = StreamReader(self)
		// try to unpack data
		return try self.unpack(stream: &reader)
	}
	
	// MARK - Unpack Internal Functions 

	/// This is the unpack function which reader
	///
	/// - Parameter stream: stream object
	/// - Returns: decoded data
	/// - Throws: an error if decoding task cannot be finished correctly due to an error
	private func unpack(stream: inout StreamReader) throws -> Any? {
		let type = try stream.readType()
		
		// Spec is defined here:
		// https://github.com/msgpack/msgpack/blob/master/spec.md#formats-bool
		switch type {
		
		// POSITIVE FIX INT
		// positive fixint	0xxxxxxx	0x00 - 0x7f
		case 0x00...0x7f:
			return Int8(type)
		
		// FIX DICTIONARY (< 16 ITEMS)
		// fixmap	1000xxxx	0x80 - 0x8f
		case 0x80...0x8f:
			let count_items = Int(type & 0xf)
			return try self.unpack(dictionary: &stream, count: count_items)
			
		// FIX ARRAY (< 16 ITEMS)
		// fixarray	1001xxxx	0x90 - 0x9f
		case 0x90...0x9f:
			let count_items = Int(type & 0xf)
			return try self.unpack(array: &stream, count: count_items)
			
		// NEGATIVE FIX NUM
		// negative fixint	111xxxxx	0xe0 - 0xff
		case 0xe0...0xff:
			return Int8( Int(type) - 256)
		
		// FIX STRING (< 16 CHARS)
		// fixstr	101xxxxx	0xa0 - 0xbf
		case 0xa0...0xbf:
			let str_length = Int(type - 0xa0)
			return try self.unpack(string: &stream, length: str_length)
		
		// NIL VALUE
		// nil	11000000	0xc0
		case 0xc0:
			return nil
			
		// BOOLEAN FALSE
		// false 	11000010	0xc2
		case 0xc2:
			return false
		
		// BOOLEAN TRUE
		// true	11000011	0xc3
		case 0xc3:
			return true
		
		// BINARY DATA 8 BIT
		// bin 8	11000100	0xc4
		case 0xc4:
			let len_data = Int(try stream.read8Bit())
			return try stream.readData(length: len_data)
			
		// BINARY DATA 16 BIT
		// bin 16	11000101	0xc5
		case 0xc5:
			let len_data = Int(try stream.read16Bit())
			return try stream.readData(length: len_data)
			
		// BINARY DATA 32 BIT
		// bin 32	11000110	0xc6
		case 0xc6:
			let len_data = Int(try stream.read32Bit())
			return try stream.readData(length: len_data)
			
		// FLOAT 32 BIT
		// float 32	11001010	0xca
		case 0xca:
			return Float(bitPattern: try stream.read32Bit())
		
		// DOUBLE
		// float 64	11001011	0xcb
		case 0xcb:
			return Double(bitPattern: try stream.read64Bit())
		
		// UNSIGNED INT 8 BIT
		// uint 8	11001100	0xcc
		case 0xcc:
			return try stream.readType()
		
		// UNSIGNED INT 16 BIT
		// uint 16	11001101	0xcd
		case 0xcd:
			let h = UInt16(try stream.read8Bit())
			let l = UInt16(try stream.read8Bit())
			return (h << 8 + l)
		
		// UNSIGNED INT 32 BIT
		// uint 32	11001110	0xce
		case 0xce:
			return try stream.read32Bit()
		
		// UNSIGNED INT 64 BIT
		// uint 64	11001111	0xcf
		case 0xcf:
			return try stream.read64Bit()
		
		// INT 8 BIT
		// int 8	11010000	0xd0
		case 0xd0:
			let value = try stream.read8Bit()
			return Int8(Int(value) - 256)
		
		// INT 16 BIT
		// int 16	11010001	0xd1
		case 0xd1:
			let h = UInt16(try stream.read8Bit())
			let l = UInt16(try stream.read8Bit())
			return Int16(bitPattern: h << 8 + l)
		
		// INT 32 BIT
		// int 32	11010010	0xd2
		case 0xd2:
			return try Int32(bitPattern: stream.read32Bit())
		
		// INT 64 BIT
		// int 64	11010011	0xd3
		case 0xd3:
			return try Int64(bitPattern: stream.read64Bit())
		
		// STRING 8 BIT LENGTH
		// str 8	11011001	0xd9
		case 0xd9:
			let len_data = Int(try stream.read8Bit())
			return try unpack(string: &stream, length: len_data)
		
		// STRING 16 BIT LENGTH
		// str 16	11011010	0xda
		case 0xda:
			let len_data = Int(try stream.read8Bit()) << 8 + Int(try stream.read8Bit())
			return try unpack(string: &stream, length: len_data)
		
		// STRING 32 BIT LENGTH
		// str 32	11011011	0xdb
		case 0xdb:
			var len_data = Int(try stream.read8Bit()) << 24
			len_data += Int(try stream.read8Bit()) << 16
			len_data += Int(try stream.read8Bit()) << 8
			len_data += Int(try stream.read8Bit())
			return try unpack(string: &stream, length: len_data)

			
		// ARRAY 16 ITEMS LENGTH
		// array 16	11011100	0xdc
		case 0xdc:
			let count_items = Int(try stream.read16Bit())
			return try unpack(array: &stream, count: count_items)
			
		// ARRAY 32 ITEMS LENGTH
		// array 32	11011101	0xdd
		case 0xdd:
			let count_items = Int(try stream.read32Bit())
			return try unpack(array: &stream, count: count_items)
			
		// DICTIONARY 16 ITEMS LENGTH
		// map 16	11011110	0xde
		case 0xde:
			let count_items = Int(try stream.read16Bit())
			return try unpack(dictionary: &stream, count: count_items)
			
		// DICTIONARY 32 ITEMS LENGTH
		// map 32	11011111	0xdf
		case 0xdf:
			let count_items = Int(try stream.read32Bit())
			return try unpack(dictionary: &stream, count: count_items)
			
		default:
			throw MsgPackError.unsupportedValue(String(format: "Type(%02x)", type))
		}
	}
	
	/// Unpack a `dictionary` sequence
	///
	/// - Parameters:
	///   - stream: input stream of data
	///   - count: number of keys in dictionary
	/// - Returns: decoded dictionary
	/// - Throws: throw an exception if failed to decoded data
	private func unpack(dictionary stream: inout StreamReader, count: Int) throws -> [AnyHashable: Any?] {
		var dict: [AnyHashable: Any?] = [:]
		for _ in 0..<count {
			guard let key = try self.unpack(stream: &stream) as? AnyHashable else {
				throw MsgPackError.unsupportedValue("Invalid dict key")
			}
			let val = try self.unpack(stream: &stream)
			dict[key] = val
		}
		return dict
	}
	
	/// Unpack an `array` sequence
	///
	/// - Parameters:
	///   - stream: input stream of data
	///   - count: number of keys in array
	/// - Returns: decoded array
	/// - Throws: throw an exception if failed to decoded data
	private func unpack(array stream: inout StreamReader, count: Int) throws -> [Any?] {
		var array: [Any?] = []
		for _ in 0..<count {
			array.append(try self.unpack(stream: &stream))
		}
		return array
	}
	
	
	/// Unpack a `string` sequence
	///
	/// - Parameters:
	///   - stream: input stream of data
	///   - length: length of data to read
	/// - Returns: decoded string
	/// - Throws: throw an exception if failed to decoded data
	private func unpack(string stream: inout StreamReader, length: Int) throws -> String {
		let data = try stream.readData(length: length)
		guard let str = String(data: data, encoding: String.Encoding.utf8) else {
			throw MsgPackError.invalidEncoding
		}
		return str
	}
	
}
